### ISSUE TITLE
MYXProtocol thrown when using large integers as prepared parameters.

### ISSUE THREAD

**Issue by nomad-software** [mysql-d/mysql-native#28]
Create a test database like this:

``` sql
DROP SCHEMA IF EXISTS `test` ;
CREATE SCHEMA IF NOT EXISTS `test` DEFAULT CHARACTER SET utf8 COLLATE utf8_bin ;
USE `test` ;

DROP TABLE IF EXISTS `test`.`user` ;

CREATE  TABLE IF NOT EXISTS `test`.`user` (
  `id` INT UNSIGNED NOT NULL AUTO_INCREMENT ,
  `dataSourceId` INT UNSIGNED NOT NULL ,
  `name` VARCHAR(64) NOT NULL ,
  `added` DATETIME NOT NULL ,
  `statusId` TINYINT UNSIGNED NOT NULL DEFAULT 1 ,
  PRIMARY KEY (`id`) ,
  UNIQUE INDEX `dataSourceId_name_UNIQUE` (`dataSourceId` ASC, `name` ASC) ,
  INDEX `fk_user_status1_idx` (`statusId` ASC) ,
  INDEX `added_INDEX` (`added` ASC) ,
  INDEX `fk_user_dataSource1_idx` (`dataSourceId` ASC))
ENGINE = InnoDB DEFAULT CHARACTER SET = utf8 COLLATE = utf8_bin;

INSERT INTO user (id, dataSourceId, name, added, statusId) VALUES
(1, 2, "test1", NOW(), 3),
(4, 5, "test2", NOW(), 6);
```

Then run the following code:

``` d
import mysql.connection;
import std.stdio;

void main(string[] args)
{
    auto connection = new Connection("localhost", "user", "password", "test");

    auto command = Command(connection);
    scope (exit) command.releaseStatement();

    command.sql  = "SELECT id, dataSourceId, LOWER(name), UNIX_TIMESTAMP(added), statusId
                      FROM user
                     WHERE id = 1
                        OR (UNIX_TIMESTAMP(added) >= (? - ?) AND UNIX_TIMESTAMP(added) <= ?)
                     LIMIT ?";

    command.prepare();

    uint baseTimeStamp    = 1371477821;
    uint cacheCutOffLimit = int.max;
    uint recordLimit      = int.max;

    command.bindParameter(baseTimeStamp, 0);
    command.bindParameter(cacheCutOffLimit, 1);
    command.bindParameter(baseTimeStamp, 2);
    command.bindParameter(recordLimit, 3);

    writefln("Retrieving results...");

    auto results = command.execPreparedResult();

    writeln("Got results."); // Doesn't reach here!
}
```

The query never completes and hangs indefinitely. This query works fine in MySQL Workbench and on the command line so i'm assuming it's an issue with this driver.

**Comment by Abscissa** [comment]
I strongly suspect the fix for #22 may have fixed this as well. Can you verify?

**Comment by nomad-software** [comment]
I think the library needs updating first:
```
/home/gary/Desktop/mysql-native/source/mysql/protocol/commands.d(288): Deprecation: module mysql.protocol.packet_helpers is not accessible here, perhaps add 'static import mysql
.protocol.packet_helpers;' 
```
After fixing that I get the following:
```
Retrieving results...
mysql.common.MySQLProtocolException@/home/gary/Desktop/mysql-native/source/mysql/result.d(129): Expected null header byte for binary result row
----------------
??:? pure bool std.exception.enforceEx!(mysql.common.MySQLProtocolException).enforceEx!(bool).enforceEx(bool, lazy immutable(char)[], immutable(char)[], ulong) [0x4f6a8c]
??:? ref mysql.result.Row mysql.result.Row.__ctor(mysql.connection.Connection, ref ubyte[], mysql.protocol.packets.ResultSetHeaders, bool) [0x4de6b3]
??:? mysql.result.ResultSet mysql.protocol.commands.Command.execPreparedResult(mysql.protocol.extra_types.ColumnSpecialization[]) [0x4e49c1]
??:? _Dmain [0x4d14b6]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x5229b2]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5228fc]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x52296e]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5228fc]
??:? _d_run_main [0x52286d]
??:? main [0x4d1cfd]
??:? __libc_start_main [0xf3d63a3f]
mysql.common.MySQLException@/home/gary/Desktop/mysql-native/source/mysql/common.d(119): Received std.socket.Socket.ERROR
----------------
??:? pure bool std.exception.enforceEx!(mysql.common.MySQLException).enforceEx!(bool).enforceEx(bool, lazy immutable(char)[], immutable(char)[], ulong) [0x4eae34]
??:? void mysql.common.MySQLSocketPhobos.read(ubyte[]) [0x4e8e16]
??:? ubyte[] mysql.connection.Connection.getPacket() [0x4e5999]
??:? ulong mysql.protocol.commands.Command.purgeResult() [0x4e36ba]
??:? void mysql.protocol.commands.Command.releaseStatement() [0x4e357e]
??:? _Dmain [0x4d14ea]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x5229b2]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5228fc]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x52296e]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5228fc]
??:? _d_run_main [0x52286d]
??:? main [0x4d1cfd]
??:? __libc_start_main [0xf3d63a3f]

shell returned 1 
```

**Comment by Abscissa** [comment]
Finally tracked this one down. Fix is on the way. (Hooray for wireshark! Even if my machine really, *really* doesn't want to allow it without sudo priveledges, grumble, grumble...)

Wireshark revealed the server was sending an error:

```
BIGINT UNSIGNED value is out of range in '(? - ?)'
```

Unfortunately, the error was being sent at a point during communications where mysql-native wasn't expecting to receive an error packet (when attempting to retrieve the first row, AFTER an initial OK packet and some column info).

It's worth noting this bug would not have occurred if packet-handling was fully encapsulated and guaranteed ONE location where packets are retrieved and checked to be an error. This is a perfect example of why the low-level internals need some serious cleanup (a current priority for mysql-native).

A fix is on the way. The timing is unfortunate, as I wanted to give the v2.1.0 release candidate one week before finalizing (five more days). Git master has undergone some internal refactorings in the meantime, which may make it difficult to cherry-pick this into v2.1.0. So there's a possibility I might just follow-up v2.1.0 with an immediate v2.1.1 including this. But this has been open for so long I'll try to get it into v2.1.0.

**Comment by Abscissa** [comment]
Included in `v2.1.0-rc3`
